[GEN][ZH] Generate git version information and print it in the Game Window title, Options Menu (and Main Menu)

xezon opened the pull request:
* Resolves #1010
* Closes #1171

This change generates information from git and makes it available in the game.

The Window Title now prints additional version information unique to this project. R stands for revision and it simply shows the commit count. We can tweak this if we want to show git tag or commit hash, but for now this appears to be sufficient. The label "By" is technically localized.

The Options Menu (and the Main Menu: "MainMenu.wnd:LabelVersion") also print additional information, including the last commit time (UTC). It also shows the EXE and INI crc's.

![versioninfo](https://github.com/user-attachments/assets/0c86c2e1-e7a9-4b16-aff3-73ec5921f8bc)

## TODO

- [x] Replicate in Generals
- [x] Log out version
- [x] Fix constness
- [x] Check EXE properties

xezon commented on resources/CMakeLists.txt:
```diff
@@ -1,6 +1,24 @@
 # Add resources (optional)
 if (MSVC_IDE)
-    add_library(resources INTERFACE
+    target_sources(resources PRIVATE
```
Yes. Because add_library now exists higher up in this file.

xezon commented on GeneralsMD/Code/GameEngine/Include/Common/version.h:
```diff
@@ -40,16 +42,26 @@ class Version
-	UnicodeString getUnicodeBuildLocation( void );	///< Return a string with the build location
-	AsciiString getAsciiBuildUser( void );					///< Return a string with the build user
-	UnicodeString getUnicodeBuildUser( void );			///< Return a string with the build user
+
```
Yes I was fixing the function constness. Do you want me to remove that?

OmniBlade replied:
No, I just was blind to the added const it seems 😄

xezon commented on GeneralsMD/Code/GameEngine/Source/Common/version.cpp:
```diff
@@ -83,19 +91,22 @@ AsciiString Version::getAsciiVersion( void )
+			version.format(TheGameText->fetch("Version:Format3").str(), m_major, m_minor, m_buildNum);
+		else
+			version.format(TheGameText->fetch("Version:Format4").str(), m_major, m_minor, m_buildNum, m_localBuildNum,
+				m_buildUser.getCharAt(0), m_buildUser.getCharAt(1));
```
Yes this Build User thing needs more work, because as you say, it is not really unique to us publishing it, if it is embedded as a plain string. I do not know what to do with it at this time.

Using Git Commit Author does not work, because that name is not stable. At least not for the CI builds. For local builds it probably would be fine.

We can use Git Commit Author, and override that with the build user if that is set (from CI).

xezon commented:
> One more thing we can do here is log out the version too.

* Fixed. Log will now print:

```
================================================================================
Generals version 1.4.601 Debug
Build date: Jul  2 2025 19:27:54
Build location: 
Build user: 
Build git revision: ~543
Build git version: ~33142610
Build git commit time: 2025-06-30 20:22:40
Build git commit author: xezon
================================================================================
```

* Build user now falls back to git commit author name if the build user is empty.

* Also fixed getAsciiVersion to be consistent with getUnicodeVersion.

* Added comments to public functions of Version class.

* Removed all the useless (void) function arguments in Version class.

xezon commented:
Rebased to Main with Debug Logging LF changes.
